### ISSUE TITLE
platformio depends on <= starlette 0.21.0, but attempts to build with Nixpkgs's 0.23.0

### ISSUE THREAD

**Issue by hacker1024** [NixOS/nixpkgs#211707]
### Steps To Reproduce
1. build `platformio`

### Build log
```
Executing pipInstallPhase
/build/source/dist /build/source
Processing ./platformio-6.1.5-py3-none-any.whl
Requirement already satisfied: ajsonrpc==1.* in /nix/store/9p9ws2gx11wqc5z0zvifswnx17na52mc-python3.10-ajsonrpc-1.2.0/lib/python3.10/site-packages (from platformio==6.1.5) (1.2.0)
Requirement already satisfied: marshmallow==3.* in /nix/store/61snn19046kbh7ammdjxlhsbajjr316k-python3.10-marshmallow-3.16.0/lib/python3.10/site-packages (from platformio==6.1.5) (3.16.0)
Requirement already satisfied: bottle==0.12.* in /nix/store/n6rz9z3z2ky6pzxj6s043l55yymkvzd4-python3.10-bottle-0.12.23/lib/python3.10/site-packages (from platformio==6.1.5) (0.12.23)
Requirement already satisfied: pyserial==3.5.* in /nix/store/jxzz20a7i2byxh0zqfk0vhh3cx8yafcz-python3.10-pyserial-3.5/lib/python3.10/site-packages (from platformio==6.1.5) (3.5)
Requirement already satisfied: click<9,>=8.0.4 in /nix/store/whih1qwwfq8a4zyzx2kdnpjvwdr9b83k-python3.10-click-8.1.3/lib/python3.10/site-packages (from platformio==6.1.5) (8.1.3)
Requirement already satisfied: semantic-version==2.10.* in /nix/store/c0rzd0dim0i5j4g6jpa5hjhajacy5g2m-python3.10-semantic-version-2.10.0/lib/python3.10/site-packages (from platformio==6.1.5) (2.10.0)
ERROR: Could not find a version that satisfies the requirement starlette<=0.21,>=0.19.1 (from platformio) (from versions: none)
ERROR: No matching distribution found for starlette<=0.21,>=0.19.1

error: builder for '/nix/store/gfbrpq67jyw8gfz5288rs3f12w7v0dx3-platformio-6.1.5.drv' failed with exit code 1
error: 1 dependencies of derivation '/nix/store/dqzgjy56m0kbda2zf7rqbb6v537gdxyx-platformio-usr-target.drv' failed to build
error: 1 dependencies of derivation '/nix/store/jzmwgkna6daw1czdzxfkb0hk3dp4kjg0-platformio-fhs.drv' failed to build
error: 1 dependencies of derivation '/nix/store/3n8l3q1p3bhryfbqw0651dw4rpj7n6i6-platformio-init.drv' failed to build
error: 1 dependencies of derivation '/nix/store/cjfsdl2ywkhzsg82clr6p95rim6rcd51-platformio.drv' failed to build
```

### Additional context
The platformio `develop` branch supports Starlette 0.23.0.
Not much changed between 0.21.0 and 0.23.0 either. The latest version may just work with 0.23.0 if setup.py is patched.

On another note:
This issue could be worked around immediately by overriding Starlette with an older version, but it's not possible to override the `platformio` derivation used in the FHS  from `core.nix`. It'd be great if there was a way to modify the derivation used in the FHS.

### Notify maintainers
@mogorman

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.86, NixOS, 23.05 (Stoat), 23.05pre443927.6dccdc45851`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.12.0`
 - channels(root): `"home-manager, nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by peterhoeg** [comment]
Here's my workaround to get it to build again:

```diff
diff --git a/pkgs/development/embedded/platformio/core.nix b/pkgs/development/embedded/platformio/core.nix
index a9b6e156b87..1d5cd60022f 100644
--- a/pkgs/development/embedded/platformio/core.nix
+++ b/pkgs/development/embedded/platformio/core.nix
@@ -42,7 +42,18 @@ with python3.pkgs; buildPythonApplication rec {
     requests
     semantic-version
     spdx-license-list-data.json
-    starlette
+    (starlette.overrideAttrs
+      (old: rec {
+        version = "0.21.0";
+        src = fetchPypi {
+          pname = "starlette";
+          inherit version;
+          hash = "sha256-sbUjBe6PfPxIzeODSW98EauJfNcRKzPZmLExfcjvkCc=";
+        };
+        postPatch = "";
+        doCheck = false;
+        doInstallCheck = false;
+      }))
     tabulate
     uvicorn
     wsproto
```

**Comment by hacker1024** [comment]
> Here's my workaround to get it to build again:
> 
> ```diff
> diff --git a/pkgs/development/embedded/platformio/core.nix b/pkgs/development/embedded/platformio/core.nix
> index a9b6e156b87..1d5cd60022f 100644
> --- a/pkgs/development/embedded/platformio/core.nix
> +++ b/pkgs/development/embedded/platformio/core.nix
> @@ -42,7 +42,18 @@ with python3.pkgs; buildPythonApplication rec {
>      requests
>      semantic-version
>      spdx-license-list-data.json
> -    starlette
> +    (starlette.overrideAttrs
> +      (old: rec {
> +        version = "0.21.0";
> +        src = fetchPypi {
> +          pname = "starlette";
> +          inherit version;
> +          hash = "sha256-sbUjBe6PfPxIzeODSW98EauJfNcRKzPZmLExfcjvkCc=";
> +        };
> +        postPatch = "";
> +        doCheck = false;
> +        doInstallCheck = false;
> +      }))
>      tabulate
>      uvicorn
>      wsproto
> ```

The [original derivation](https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/starlette/default.nix) actually fetches sources from GitHub instead of PyPI.

I built it like this, and didn't need to disable any tests - not sure if that's related?

```nix
starlette.overrideAttrs ({ pname, ... }: rec {
  version = "0.21.0";
  src = fetchFromGitHub {
    owner = "encode";
    repo = pname;
    rev = version;
    hash = "sha256-wZtlKCD7eE4sCKfhFgWXywACAes8uo4HXkicJAMcgqk=";
  };
})
```

**Comment by peterhoeg** [comment]
> The original derivation actually fetches sources from GitHub instead of PyPI.
> I built it like this, and didn’t need to disable any tests - not sure if that’s related?

Often the tests are not available in the distribution bundle from pypi, so that makes sense. Yours is better!

**Comment by peterhoeg** [comment]
Do you want to prep a PR?

**Comment by gador** [comment]
solved by 62b9dc3abb57084d9c50806c34bbee8f098fad17
